standard/statements.md

```diff
@@ -1008,7 +1043,7 @@ The target of a `goto` *identifier* statement is the labeled statement with the
 > ```
 > a `goto` statement is used to transfer control out of a nested scope. *end note*
 
-The target of a `goto case` statement is the statement list in the immediately enclosing `switch` statement ([§13.8.3](statements.md#1383-the-switch-statement)) which contains a`case` label with the given constant value. If the `goto case` statement is not enclosed by a `switch` statement, if the *constant_expression* is not implicitly convertible ([§11.2](conversions.md#112-implicit-conversions)) to the governing type of the nearest enclosing `switch` statement, or if the nearest enclosing `switch` statement does not contain a `case` label with the given constant value, a compile-time error occurs.
+The target of a `goto case` statement is the statement list in the immediately enclosing `switch` statement ([§13.8.3](statements.md#1383-the-switch-statement)) which contains a `case` label with the given *constant_pattern*. If the `goto case` statement is not enclosed by a `switch` statement or if the nearest enclosing `switch` statement does not contain a `case` label with the given *constant_pattern*, a compile-time error occurs.
```

Comment:
With the given *constant_pattern*?  Do we have a concept of when two constant patterns are the same?

Reply:
There is also a requirement that the target case label must not have a `when` clause.

Reply:
I'll take a look.

Reply:
@gafter Re your Q, "Do we have a concept of when two constant patterns are the same?" In the new section "Constant patterns," we have the following: "Given a pattern input value *e* and a *constant_expression* *c*, if *e* and *c* have integral types, the pattern is considered matched if the result of the expression `e == c` is `true`. Otherwise, the pattern is considered matched if `object.Equals(e, c)` returns `true`. In this case it is a compile-time error if the static type of *e* is not pattern compatible (§declaration-pattern-new-clause) with the type of the constant." Does that address your concern?

In the context of a switch, `goto case 0;`, `goto case 10 - 10;`, and `goto case 5 + 2 - 7;` all match `case 0:`.

Reply:
@gafter  Re your comment, "There is also a requirement that the target case label must not have a when clause.", I have confirmed that. Given the case label `case 10 when flag == true:`, `goto case 10;` results in CS0159 "No such label 'case 10' within the scope of the goto statement."

I propose adding the following:
> It is a compile-time error for *constant_pattern* to target a *switch_label* containing a *case_guard*.

Reply:
No, that doesn't quite do it.  The text "The target of a `goto case` statement is the statement list in the immediately enclosing `switch` statement ([§13.8.3](statements.md#1383-the-switch-statement)) which contains a `case` label with the given *constant_pattern*." needs to be modified, because there may be more than one such target.

```csharp
int i = 12;
switch (i)
{
    case 1 when "".Length == 0:
        break;
    case 1:
        break;
    case 2:
        goto case 1;
}
```